File: structures-js/structures-api/src/api/idl/decorators/EntityServiceDecoratorsConfig.ts
import {EntityServiceDecorator} from '@/api/idl/decorators/EntityServiceDecorator'
import {IEntityService} from '@/api/IEntityService'

// Helper type to determine if a type is a function
type IfFunction<T, U> = T extends (...args: any[]) => any ? U : never;

// Extracting only method names
type MethodsOf<T> = {
    [K in keyof T]: IfFunction<T[K], K>;
}[keyof T];

// List of methods to exclude
type ExcludedMethods = 'namedQuery' | 'namedQueryPage';

// Filtered methods from IEntityService
type FilteredMethods<T> = Exclude<MethodsOf<T>, ExcludedMethods>;

// Defining the DecoratedFunction type based on IEntityService methods, excluding some, and adding a few more
export type DecoratedFunction = FilteredMethods<IEntityService<any>> | 'allCreate' | 'allRead' | 'allUpdate' | 'allDelete';

/**
 * Configuration for the {@link EntityServiceDecoratorsDecorator}
 * This is a map of method names to an array of decorators that should be applied to that method
 */
export type EntityServiceDecoratorsConfig = Partial<Record<DecoratedFunction, EntityServiceDecorator[]>>

File: structures-core/src/main/java/org/kinotic/structures/api/domain/idl/decorators/EntityServiceDecoratorsConfig.java
package org.kinotic.structures.api.domain.idl.decorators;

import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.kinotic.structures.internal.api.services.EntityService;

import java.util.List;

/**
 * Configuration for the {@link EntityServiceDecoratorsDecorator}
 * This maps {@link EntityService} method names to decorators that should be applied to the method
 */
@Getter
@Setter
@Accessors(chain = true)
public class EntityServiceDecoratorsConfig{

    /**
     * The following map crud operations to decorators
     * These will be applied to the corresponding method type in the {@link EntityService}
     * i.e. create will be applied to the save method, read will be applied to count, findAll, findById, ect
     */
    private List<EntityServiceDecorator> allCreate = List.of();
    private List<EntityServiceDecorator> allRead = List.of();
    private List<EntityServiceDecorator> allUpdate = List.of();
    private List<EntityServiceDecorator> allDelete = List.of();

    /**
     * The following map individual methods to decorators
     * These will be applied to the corresponding method in the {@link EntityService}
     * i.e. bulkSave will be applied to the bulkSave method
     */
    private List<EntityServiceDecorator> bulkSave = List.of();
    private List<EntityServiceDecorator> bulkUpdate = List.of();
    private List<EntityServiceDecorator> count = List.of();
    private List<EntityServiceDecorator> countByQuery = List.of();
    private List<EntityServiceDecorator> deleteById = List.of();
    private List<EntityServiceDecorator> deleteByQuery = List.of();
    private List<EntityServiceDecorator> findAll = List.of();
    private List<EntityServiceDecorator> findById = List.of();
    private List<EntityServiceDecorator> findByIds = List.of();
    private List<EntityServiceDecorator> syncIndex = List.of();
    private List<EntityServiceDecorator> save = List.of();
    private List<EntityServiceDecorator> search = List.of();
    private List<EntityServiceDecorator> update = List.of();
}
